feat(bullmq): add bullmq (>=2) instrumentation

unflxw opened the pull request:
Add an instrumentation for BullMQ. This commit adds the already existing instrumentation at [appsignal/opentelemetry-instrumentation-bullmq](https://github.com/appsignal/opentelemetry-instrumentation-bullmq) to this repository, with some minor changes, removing under-tested functionality in order to reduce the maintenance surface.

The initial work for this instrumentation was done by @jenniferplusplus at [jenniferplusplus/opentelemetry-instrumentation-bullmq](https://github.com/jenniferplusplus/opentelemetry-instrumentation-bullmq), of which both appsignal/opentelemetry-instrumentation-bullmq and the instrumentation in this commit are forks.

The instrumentation aims to comply with the Messaging Semantic Conventions (v0.24.0) whenever possible. Due to the latest published version of the `@opentelemetry/semantic-conventions` package not fully complying with the semantic conventions, when in conflict, the instrumentation chooses to follow the implemented shared package instead of the contents of the document.

jenniferplusplus commented:
It's good to see this happening. I'm glad the instrumentation has been useful to people, but I'm also glad to be relieved of it. It's not a stack I use anymore.

I'll be happy to mark my npm package as deprecated once this gets merged and published.

unflxw commented:
Thank you!

The same goes for `@appsignal/opentelemetry-instrumentation-bullmq`, we would deprecate it and move people over to the contrib-based instrumentation.

pichlermarc commented:
Hi @unflxw. We've recently added [new contributing guidelines for new instrumentation libraries](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#new-instrumentation).

Is there a specific reason why you'd want `@appsignal/opentelemetry-instrumentation-bullmq` here in this repo as opposed to hosting and maintaining it in a seperate repo, ideally closer to the community that knows and maintains BullMQ?

unflxw commented:
@pichlermarc Thanks for letting me know! I had not seen these guidelines before.

Our belief is that it is valuable for this instrumentation to be visible to the OpenTelemetry community, as well as for its maintenance to be shared with the community at large. Nonetheless, I have added myself as the maintainer for this instrumentation in the component owners file.

We're not BullMQ maintainers ourselves. We run an OpenTelemetry-based APM, and we took on the task of improving this instrumentation in order to be able to offer it to our customers. So, while we can host it on a public repository of our own, as we currently do, it will not be a repository owned by the BullMQ maintainers. The BullMQ maintainer team has [stated that they're not interested](https://github.com/taskforcesh/bullmq/issues/2048#issuecomment-1672825524) in maintaining the OpenTelemetry instrumentation.

The instrumentation, while popular enough to have [tens of thousands of weekly downloads](https://www.npmjs.com/package/@jenniferplusplus/opentelemetry-instrumentation-bullmq), was [not under active maintenance](https://github.com/jenniferplusplus/opentelemetry-instrumentation-bullmq/issues/20#issuecomment-2059287155). Part of why we forked it was to avoid imposing additional maintenance workload on @jenniferplusplus, as we implemented changes to it. We would hope to prevent further fragmentation of the instrumentation ecosystem and the work on it by sponsoring it under the OpenTelemetry banner.

unflxw commented:
Please let us know if there's anything we can do to help move this forward.
